File: app/src/main/java/com/jkristian/ioio/scope/LineSet.java
package com.jkristian.ioio.scope;

import android.view.View;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Deque;
import java.util.Iterator;

/**
 * Converts a sequence of samples to a sequence of lines.
 */
class LineSet {

    static final int STROKE = 4;
    private static final int HALF_STROKE = 2;
    private static final int SCALE = 16; // pixels/second
    private static final long SEC = 1000000000L;

    final int width;
    final int height;
    private final long startTime;
    private final Collection<Sample> samples;
    private final long now = System.nanoTime();
    private Collection<Line> lines = null;

    LineSet(View chart, Deque<Sample> samples) {
        width = chart.getWidth();
        height = chart.getHeight();
        startTime = now - (width * SEC / SCALE);
        this.samples = recent(samples);
    }

    boolean isEmpty() {
        return samples.isEmpty();
    }

    float[] toPoints() {
        if (lines == null) {
            computeLines();
        }
        float[] points = new float[lines.size() * 4];
        int p = 0;
        for (Line line : lines) {
            points[p++] = line.fromX;
            points[p++] = line.fromY;
            points[p++] = line.toX;
            points[p++] = line.toY;
        }
        return points;
    }

    private void computeLines() {
        lines = new ArrayList<Line>(samples.size() + 1);
        if (samples.isEmpty()) {
            return;
        }
        Sample last = null;
        for (Sample sample : samples) {
            if (last != null) {
                addLineFrom(last, sample.time);
            }
            last = sample;
        }
        addLineFrom(last, now);
    }

    private void addLineFrom(Sample start, long endTime) {
        if (start.time <= endTime) {
            int lastX = Math.max(0, (int) ((start.time - startTime) * SCALE / SEC));
            int x = Math.min(width - 1, (int) ((endTime - startTime) * SCALE / SEC));
            int y = HALF_STROKE + Math.round((1.0f - start.value) * (height - 1 - HALF_STROKE));
            lines.add(new Line(lastX, y, x, y));
        }
    }

    private Collection<Sample> recent(Deque<Sample> from) {
        while (from.size() > 1 && getSecond(from).time < startTime) {
            from.removeFirst();
        }
        return new ArrayList<>(from);
    }

    private static <T> T getSecond(Iterable<T> from) {
        Iterator<T> i = from.iterator();
        i.next();
        return i.next();
    }
}

File: app/src/main/java/com/jkristian/ioio/scope/LineSet.kt
package com.jkristian.ioio.scope

import android.view.View

import java.util.ArrayList

/**
 * Converts a sequence of samples to a sequence of lines. Not thread safe.
 */
internal class LineSet(chart: View) {

    val width: Int
    val height: Int
    private val now = System.nanoTime()
    private val startTime: Long

    init {
        width = chart.width
        height = chart.height
        startTime = now - width * SEC / SCALE
    }

    fun toPoints(samples: SampleSet): FloatArray {
        val lines = toLines(samples.getRecent(startTime))
        val points = FloatArray(lines.size * 4)
        var p = 0
        for (line in lines) {
            points[p++] = line.fromX.toFloat()
            points[p++] = line.fromY.toFloat()
            points[p++] = line.toX.toFloat()
            points[p++] = line.toY.toFloat()
        }
        return points
    }

    private fun toLines(samples: Collection<Sample>): Collection<Line> {
        val lines = ArrayList<Line>(if (samples == null) 1 else samples.size + 1)
        if (!samples.isEmpty()) {
            var last: Sample? = null
            for (sample in samples) {
                if (last != null) {
                    addLineFrom(lines, last, sample.time)
                }
                last = sample
            }
            addLineFrom(lines, last!!, now)
        }
        return lines
    }

    private fun addLineFrom(lines: MutableCollection<Line>, start: Sample, endTime: Long) {
        if (start.time <= endTime) {
            val fromX = Math.max(0, ((start.time - startTime) * SCALE / SEC).toInt())
            val toX = Math.min(width - 1, ((endTime - startTime) * SCALE / SEC).toInt())
            val y = HALF_STROKE + Math.round((1.0f - start.value) * (height - 1 - HALF_STROKE))
            lines.add(Line(fromX, y, toX, y))
        }
    }

    companion object {

        val STROKE = 4
        private val HALF_STROKE = 2
        private val SCALE = 16 // pixels/second
        private val SEC = 1000000000L
    }
}